include/geode/geometry/detail/aabb_impl.h

```diff
@@ -391,4 +409,114 @@ namespace geode
                 action );
         }
     }
+
+    template < index_t dimension >
+    template < typename ACTION >
+    void AABBTree< dimension >::Impl::ray_trace_intersect_recursive(
+        const InfiniteLine< dimension >& ray,
+        index_t node_index,
+        index_t element_begin,
+        index_t element_end,
+        ACTION& action ) const
+    {
+        OPENGEODE_ASSERT( node_index < tree_.size(), "Node out of tree range" );
+        OPENGEODE_ASSERT(
+            element_begin != element_end, "No iteration allowed start == end" );
+
+        // Prune sub-tree that does not have intersection
+        if( !ray_box_intersection( ray, node( node_index ) ) )
+        {
+            return;
+        }
+
+        // Leaf case
+        if( is_leaf( element_begin, element_end ) )
+        {
+            // @todo Check if the box is not intersecting itself
+            action( mapping_morton_[element_begin] );
+            return;
+        }
+
+        index_t box_middle, child_left, child_right;
+        get_recursive_iterators( node_index, element_begin, element_end,
+            box_middle, child_left, child_right );
+
+        ray_trace_intersect_recursive< ACTION >(
+            ray, child_left, element_begin, box_middle, action );
+        ray_trace_intersect_recursive< ACTION >(
+            ray, child_right, box_middle, element_end, action );
+    }
+
+    template < index_t dimension >
+    bool ray_box_intersection( const InfiniteLine< dimension >& ray,
+        const BoundingBox< dimension >& box )
+    {
+        const auto box_center = ( box.min() + box.max() ) / 2.;
+        const auto box_extent = ( box.max() - box.min() ) / 2.;
+
+        // Transform the ray to the aligned-box coordinate system.
+        const auto ray_origin = ray.origin() - box_center;
+        const auto rayDirection = ray.direction();
```

Comment:
```suggestion
        const auto& ray_direction = ray.direction();
```
Is this variable necessary?